### PR TITLE
Background image: Add backgroundSize and repeat features

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Part of: https://github.com/WordPress/gutenberg/issues/54336

Explore adding a `backgroundSize` and `backgroundRepeat` set of controls for use with the background image block support. For the Group block, this allows users to set the background image to "Cover" (default), "Contain", or a custom size, which exposes repeating backgrounds automatically.

After some exploration, background repeat is included alongside backgroundSize as users will nearly always need to interact with background size when adjusting background repeat, so it's simpler and easier to keep them combined rather than separate.

## Why?

* Allow users to have repeating backgrounds, with control of the background image's size for repeating.
* Allow users to have contained background images, that fills one dimension and repeat along the other.

## How?

* Add a `backgroundSize` property and set of controls to the background block support
* Within these controls include a toggle for toggling repeating backgrounds on and off
* Opt the Group block in to this property (but have it hidden by default)
* Treat no value (undefined) as `cover`, with options for `contain` and custom sizes
* An empty string (switching to "Fixed") will default the output to the equivalent of `auto`, allowing repeating backgrounds
* Add a toggle for toggling repeat on and off. For the `contain` option, default to centering the background position. Note: background position controls will be explored in a separate PR.
* Note: global styles support is _not_ included in this PR, as the background image support will need separate work for global styles.

## To-do

* [x] Rename Fixed to Repeat, make clicking the toggle set the size to auto
* [x] Add controls to adjust the background size (so you can make it smaller or larger)
* [x] Ensure the controls are hidden when they should be
* [x] Hide these controls by default?
* [x] Add background repeat toggle
* [x] Rename Size to Fixed as the toggle option
* [x] Add centre positioning by default for the contained option
* [x] In the label and dropdown, rename to Size

## Testing Instructions

1. With TT4 theme active, add a Group block to a post, page, or template
2. Give the Group block a background image in the inspector controls
3. Click the ellipsis button to expose the background size controls
4. Have a play with the different options and try out setting a custom size value, and resetting values
5. How does it feel to use?

### Screenshot

image

### Screengrab

https://github.com/WordPress/gutenberg/assets/14988353/9337346e-7acc-444b-8e9f-2c095321f8a3

### CODE REVIEW

**Comment by andrewserong** [comment]
Okay, I think this PR is ready for early feedback on how it feels to use and what the controls might look like. This PR adds options for `cover` (default), `contain`, and custom sizes for background images, which allows repeating background images with control over the size of that background image.

I'm curious if folks think if we also need a toggle for control over repeat at this stage? I.e. on / off / x / y options, or if it's decent enough to start out with implicitly having repeating backgrounds whenever the size is not `cover` 🤔

File: packages/block-editor/src/hooks/background.js [hunk 1]
```diff
@@ -302,18 +340,150 @@ function BackgroundImagePanelItem( { clientId, setAttributes } ) {
...
+				/>
+				<ToggleGroupControlOption
+					key={ 'size' }
+					value={ 'auto' }
+					label={ __( 'Size' ) }
+				/>
+			</ToggleGroupControl>
+			{ value !== undefined &&
+			value !== 'cover' &&
+			value !== 'contain' ? (
```
Comment (ramonjd) [inline packages/block-editor/src/hooks/background.js]: Is there a reason not to use `value === 'size'`?

File: packages/block-editor/src/hooks/background.js [hunk 2]
```diff
@@ -302,18 +340,150 @@ function BackgroundImagePanelItem( { clientId, setAttributes } ) {
...
+				<UnitControl
+					size={ '__unstable-large' }
```
Comment (ramonjd) [inline packages/block-editor/src/hooks/background.js:492]: Definitely a 'later' thing, but do you reckon it'd be useful follow up to have width and height controls? E.g., be able to produce values like `background-size: 50px 100px;`

File: packages/block-editor/src/hooks/background.js [hunk 3]
```diff
@@ -302,18 +340,150 @@ function BackgroundImagePanelItem( { clientId, setAttributes } ) {
...
 }
 
+function backgroundSizeHelpText( value ) {
+	if ( value === 'cover' || value === undefined ) {
+		return __( 'Stretch image to cover the block.' );
+	}
+	if ( value === 'contain' ) {
+		return __( 'Image is contained within the block.' );
+	}
+	return __( 'Repeat the image, and set a fixed size.' );
```
Comment (ramonjd) [inline packages/block-editor/src/hooks/background.js]: I don't know if this matters, but should we be clear that the value affects the `width` only? (height is `auto`)

I believe I'm overthinking it here.

**Review by ramonjd — COMMENTED** [review]
Working excellently for me so far.

> Should we be able to toggle "repeat" off? (i.e. options for repeat of "off", "x", "y", and "both"?)

I think it would be useful to be able to toggle repeat on and off for starters, something that would play well with `background-position` if it's ever implemented.

File: packages/block-editor/src/hooks/background.js (hunk 3 above)
Comment (ramonjd) [reply packages/block-editor/src/hooks/background.js]: I forgot to add to my comments, I know this is WIP so feel free to ignore all my "observations" 😄

File: packages/block-editor/src/hooks/background.js (hunk 2 above)
Comment (andrewserong) [reply packages/block-editor/src/hooks/background.js:492]: Yes — I was imagining in follow-ups we could introduce a split control. I imagine in most cases folks will want to only use a single size value so that the image's proportions are maintained.

File: packages/block-editor/src/hooks/background.js (hunk 1 above)
Comment (andrewserong) [reply packages/block-editor/src/hooks/background.js]: Yes, we never actually set the value to `size` (size is just the current label and key for the control). In order for the active `ToggleGroupControl` to match the expected value, we need to do some mapping here between what the real value is, and what the group control thinks the value is. So the logic is:

* Any value that will either be output as `auto` or a fixed size, we treat as `auto` so that the Size toggle is highlighted. That toggle button's value needs to be `auto` so that clicking on it sets the correct value.
* We then fall back to `value` to be used to determine which control is active.
* Finally, use the default fallback of `cover` in case anything has slipped through the conditions (this is mostly so that `null` or `undefined` is treated as `cover` for the initial state), though it shouldn't _really_ ever get hit once a user has selected a value.

TL;DR, we use 'size' as a label, but `auto` or a fixed value as the real value, and need to treat that as `auto` for the right button to be highlighted.

Hope that all makes sense, let me know if you think it can be simplified, though!

File: packages/block-editor/src/hooks/background.js (hunk 3 above)
Comment (andrewserong) [reply packages/block-editor/src/hooks/background.js]: Oh, good point! I've been using squarish images in my testing so didn't notice. We can change the help text to say `fixed width` instead 👍

File: packages/block-editor/src/hooks/background.js (hunk 1 above)
Comment (ramonjd) [reply packages/block-editor/src/hooks/background.js]: Thanks for the explainer!

**Comment by andrewserong** [comment]
Thanks for taking this for a spin @ramonjd! 🙇  

> I think it would be useful to be able to toggle repeat on and off for starters, something that would play well with background-position if it's ever implemented.

Totally. If / when this PR looks viable, I was imagining doing `background-position` as the next control to implement 👍

The main challenge I've had so far is trying to work out where the toggle might appear / how it would look! This PR seems to work pretty well using a `ToggleGroupControl` for the different sizes. I wasn't too sure how the repeat toggle should work or look. Some questions / ideas:

* Do we have it as a boolean toggle that defaults to on?
* Is it displayed when you activate backgroundSize, or is it its own control in the ToolsPanel menu?
* Is starting with a boolean toggle good enough, or do we want control over x and y repeat at the outset?

Very happy for any and all ideas, and I can see about hacking them in 😄

**Comment by ramonjd** [comment]
> Do we have it as a boolean toggle that defaults to on?
> Is it displayed when you activate backgroundSize, or is it its own control in the ToolsPanel menu?

That was my first thought, though someone more qualified might have a better idea 😄 

But then I remembered `background-repeat` is not tied in any way to size and can stand on its own. So maybe it's own control? 

If folks don't like the idea of a toggle control, then perhaps something like text decoration could work, with "reset" and "repeat" buttons, e.g., 

Screenshot 2023-12-19 at 10 38 43 am

> Is starting with a boolean toggle good enough, or do we want control over x and y repeat at the outset?

Good question 🤔  Without going into the details, a boolean _sounds_ easier assuming it doesn't present any backwards compat complications should x and y become a thing later? Or could it work similar to axial spacing, e.g., toggle between 1 and 2 inputs?

**Comment by andrewserong** [comment]
Thanks for the thoughts, Ramon, much appreciated! I'll have a go putting it in as a single toggle, and then we can have a play with it and see how it feels to use 🙂

File: packages/block-editor/src/hooks/background.js (hunk 3 above)
Comment (andrewserong) [reply packages/block-editor/src/hooks/background.js]: Updated the help text in https://github.com/WordPress/gutenberg/pull/57005/commits/53e38e343c415bd325681450f646a6891b823bae

**Comment by SaxonF** [comment]
Nice work @andrewserong . A couple suggestions:

- "Size" could be "Fixed" to align with width/height setting
- "Image is contained within the block." to "Resize image to fit without cropping"
- Contain is centre positioned by default
- "Background size" to "Size"
- Add repeat tool that uses toggle that's disabled by default "Repeat image"

Once we get around to adding position we might want to condense controls.

**Comment by andrewserong** [comment]
Nice one, thanks Saxon! 🙇

**Comment by andrewserong** [comment]
Update: I've worked through and added in all the feedback from Saxon in https://github.com/WordPress/gutenberg/pull/57005#issuecomment-1862068914, so this should be ready for review now!

The main new things are:

* I've added in a toggle control for switching on and off repeating backgrounds. At first I tried this as a separate ToolsPanelItem, but it quickly became obvious that it's a real pain to have it a separate control as you'll nearly always have to interact with size at the same time as repeat, so it's better to have them combined in the one control.
* I've added in behaviour so that switching to `contain` toggles the repeating background off by default, and also outputs a centered background position. In this PR folks can toggle the repeating background, but I'll leave controls for the background position to another PR.

I'll be AFK after tomorrow, so happy to continue on with this PR in the new year if the current shape isn't quite ready for landing 🙂

**Comment by ramonjd** [comment]
Working well in all browsers for me. 🎉 

A couple of things struck me:

1. Should we hide the "Repeat" toggle for the Contain selection?
2. Setting a fixed size re-selects "repeat" if it's toggled off. Is that intentional?

https://github.com/WordPress/gutenberg/assets/6458278/81790032-2a40-4889-9b32-8a172dcffb26

**Comment by andrewserong** [comment]
Thanks for taking this for a spin @ramonjd! I've done a couple of updates in https://github.com/WordPress/gutenberg/pull/57005/commits/10ac7f8bb309f11ca5030cac2364d00236710382:

* Hide the `repeat` toggle when using the `cover` size (we still want it for `contain` so that repeating background textures like rock textures that should fit all the way horizontally can be repeated vertically, for example)
* Updating the fixes size should now no longer affect the value of "repeat". However, switching between the toggles, if you go from "contain" to "fixed" it should default to "on" for repeat

Here's how it's looking now:

https://github.com/WordPress/gutenberg/assets/14988353/2ea8b33e-2494-42eb-9a7d-3a9fbfc06730

File: packages/style-engine/class-wp-style-engine.php [hunk 4]
```diff
@@ -44,14 +44,26 @@ final class WP_Style_Engine {
...
+				),
+				'path'          => array( 'background', 'backgroundPosition' ),
+			),
+			'backgroundRepeat'   => array(
+				'property_keys' => array(
+					'default' => 'background-repeat',
+				),
+				'path'          => array( 'background', 'backgroundRepeat' ),
+			),
+			'backgroundSize'     => array(
```
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine.php:66]: Would be good to update the tests in the Core patch 👍🏻

File: packages/style-engine/src/styles/background/index.ts [hunk 5]
```diff
@@ -40,4 +40,53 @@ const backgroundImage = {
...
+		// If background size is set to contain, but no position is set, default to center.
+		if (
+			_backgroundSize === 'contain' &&
+			_backgroundPosition === undefined
+		) {
+			styleRules.push( {
+				selector: options.selector,
+				key: 'backgroundPosition',
+				value: 'center',
+			} );
```
Comment (ramonjd) [inline packages/style-engine/src/styles/background/index.ts:85]: Can be a follow up, but having a couple of tests might help our future selves 😄

File: packages/block-editor/src/hooks/background.js [hunk 6]
```diff
@@ -302,18 +346,200 @@ function BackgroundImagePanelItem( { clientId, setAttributes } ) {
...
+			sizeValue !== 'cover' &&
+			sizeValue !== 'contain' ) ||
+		sizeValue === ''
+			? 'auto'
+			: sizeValue || 'cover';
+
+	// If the current value is `cover` and the repeat value is `undefined`, then
+	// the toggle should be unchecked as the default state. Otherwise, the toggle
+	// should reflect the current repeat value.
+	const repeatCheckedValue =
```
Comment (ramonjd) [inline packages/block-editor/src/hooks/background.js:388]: Because it's a bool, also possible to do something like:

```
const repeatCheckedValue = repeatValue !== 'no-repeat' && !(currentValueForToggle === 'cover' && repeatValue === undefined);
```

Just an observation. No need to change if it works! 😄

**Review by ramonjd — APPROVED** [review]
Thanks for updating the PR so quickly. I can confirm that the controls now work as I'd expect.

Styles appear as they should in editor and frontend on Firefox, Chrome, Safari and Edge.

Enabling `appearance-tools` on a classic theme enables the controls.

Just left some optional comments. Would be nice to have tests to complement existing style engine rules in a follow up / in the Core patch for PHP.

NICE!

**Comment by andrewserong** [comment]
Thanks so much @ramonjd! 🙇

I'll leave this PR open for now since it's only had a brief look from design, and we might get a bit more feedback overnight potentially. If it all looks okay, I might merge this in tomorrow — if not, I'll look to landing this in the new year along with follow-ups for adding some more tests as you suggest 👍

Thanks again!

**Comment by jasmussen** [comment]
Looks pretty good:

background

There's a bit of an awkwardness in that you can clear the background image, but still have the repeating/size properties. Not sure if we should clear that also, mainly because it does nothing unless there's a background image. 

Seems pretty safe to add this, though, as it's a non-default ellipsis tool.

**Comment by jameskoster** [comment]
This seems okay for now, but a couple of considerations:

1. The controls feel a bit heavy/overemphasised given they're pretty much set-and-forget. 
2. This implementation won't work when we add support for multiple background images.

I think both issues could be solved by somehow 'attaching' the size settings to the background image control. That would also solve the issue about the controls being available even when there's no image set). A rough mockup based on the color picker popover:

Background size

This also creates a space for additional controls (e.g. filters) in the future.

**Comment by fabiankaegy** [comment]
@jameskoster I really like the design you are showing here. 

Something that I think would be great to add in this design exploration however is how we could deal with a focal point picker. 

One of the great things we can do with background images is setting the `background-position` and especially with the new size controls added here, being able to control the focal point would be a logical next step. 

However, the design that you have shown would limit us there because of the way the replace control is overlayed on the image.

**Comment by jameskoster** [comment]
Oh good point. Fortunately it should be easy enough to move the Replace action. Potential up a level to the Inspector:

Screenshot 2023-12-21 at 11 55 33

**Comment by ramonjd** [comment]
I think we have a good foundation to build upon.

Would be good to follow this PR up with folks' design suggestions. Might make the PR reviews easier too. 

🚢

**Comment by andrewserong** [comment]
Thanks for the feedback and ideas, folks! 🙇

> Seems pretty safe to add this, though, as it's a non-default ellipsis tool.
> This seems okay for now

Since there's no apparent blockers and this feature is hidden by default, I'll merge this in with its current form, and will look into a redesign / follow-ups early next year when we tackle the background position controls.

> Would be good to follow this PR up with folks' design suggestions. Might make the PR reviews easier too.

Yes — I reckon it'll be easier to review in follow-ups once the feature is already in place, as then we're just looking at the UI rather than all the other pieces of the support 👍

> There's a bit of an awkwardness in that you can clear the background image, but still have the repeating/size properties. Not sure if we should clear that also, mainly because it does nothing unless there's a background image.

It'd be a pretty simple follow-up to clear out the other controls when resetting the background image, so we can totally do that. One (potential) reason to keep them a little separate is so that a user can clear out / reset / swap out images with ease while they're working on something without losing their other settings. The use case there is a user cycling through different images and playing around different options where it might be jarring if you accidentally cleared everything out when removing an image. I don't feel strongly either way, though 😄

> @jameskoster I really like the design you are showing here.

Same, that looks great! It also recognises that once you have an image set, you're probably going to want to access all the available controls in one place. I'd love to have a go at implementing that in the new year — this PR already saves the attributes in the right place, so we should be able to safely refactor how we're presenting the tools without affecting any of the saved data.

One question in looking at those designs: what triggers the two different popovers? (The replace dropdown versus the popover with the background image and other controls). Is the whole item no longer a single button, but technically two different buttons, with the text and circle on the left opening the main popover, and the icon on the right opening the replace action? We might need to take care there with keyboard controls and how we navigate in and out of those different dropdowns/popovers, as we're effectively changing how the button is interacted with between its empty and active states (going from a single button to a split set of two buttons) 🤔

Feel free to share any further info or ideas either here or in https://github.com/WordPress/gutenberg/issues/54336, and I'll dig in next year!

**Comment by jameskoster** [comment]
> One question in looking at those designs: what triggers the two different popovers?

I suppose it would function a bit like the rows in List View, where you can click a row to select but the ellipsis icon is also interactive: 

Screenshot 2023-12-22 at 13 11 54

I think this pattern could have practical uses elsewhere, e.g.: https://github.com/WordPress/gutenberg/issues/41866

But I recognise it may be finicky to implement. Putting those controls in the popover could work too.

**Comment by andrewserong** [comment]
Backported in: https://core.trac.wordpress.org/ticket/60175

**Comment by paaljoachim** [comment]
A client noticed that the focus on the background image in a stack was from the top going downward.
She thought it strange that the focus of the image was not in the center.

**Comment by getdave** [comment]
Updated on [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) to denote complete. Thank you for working on this.

**Comment by andrewserong** [comment]
> A client noticed that the focus on the background image in a stack was from the top going downward.
She thought it strange that the focus of the image was not in the center.

Thanks Paal — this feature added in background size, but not background position controls. For now, if folks need control over background position, they may be better off using the Cover block. My plan is to add in background position control, too, but that feature might not be ready in time for WP 6.5.

**Comment by andrewserong** [comment]
Draft dev note:

---

## Background image block support additional features for size, repeat, and position

In WordPress 6.5 the background image block support receives new controls related to the size of the background image: size, repeat, and position.

For blocks that use the background image block support (currently the Group block is the only core block to use this support), these optional controls can now be displayed. They are hidden by default behind the tools panel menu in the Background controls in the inspector sidebar.

### How to add background size support to a theme

There are two ways to add support for `backgroundSize` and the related size, repeat and position controls to a block theme. The simplest is to opt in to the `appearanceTools` setting, which automatically enables a number of design tools ([read more in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls)).

For themes that wish to have more granular control over which UI tools are enabled, the `backgroundSize` support can be opted into by setting `settings.background.backgroundSize` to `true` in `theme.json`. For example:

```
{
	"settings": {
		"background": {
			"backgroundImage": true,
			"backgroundSize": true
```

Note that as of WP 6.4 and WP 6.5, the `backgroundImage`, `backgroundSize` and related supports are only available at the individual block level, and not in global styles or at the site root. These features will be explored in subsequent releases, with progress tracked in this Github issue: https://github.com/WordPress/gutenberg/issues/54336.
